### ISSUE TITLE
Feature Request: Provide a means for clients to define custom Xcode schemes

### ISSUE THREAD

**Issue by cgrindel** [buildbuddy-io/rules_xcodeproj#573]
Today, `rules_xcodeproj` generates an Xcode scheme for each target. It would be useful for clients to be able to define and configure the Xcode schemes as one can do in the Xcode user interface. 

## Suggested Features
- Specify a custom name
- Specify the targets to build in each scheme. 
- Identify the test targets in the build list and use them as the testables.
- Specify the launch target.

## Tasks

- [x] #592
- [x] #612
- [x] #626 
- [x] #630
- [x] #637 
- [x] #646 
- [x] #662 
- [x] #664 
- [x] #656 
- [x] #733
- [x] Generate custom Xcode schemes.
  - [x] #748 
  - [x] #801
  - [x] #807 
  - [x] #803
  - [x] #808 
  - [x] #809
  - [x] #844
  - [x] Remove or adjust top-level target logic: [top-level inference](https://github.com/buildbuddy-io/rules_xcodeproj/issues/573#issuecomment-1197341285), [isTopLevel logic](https://github.com/buildbuddy-io/rules_xcodeproj/issues/573#issuecomment-1197346855)
    - [x] #846
    - [x] #855 
    - [x] #859
    - [x] #865
    - [x] #867 
  - [x] Update "best" host logic selection in `XCSchemeInfo.TargetInfo` when the best host is not a top-level target. (see GH573 comment)
    - [x] #889 
    - [x] #890 
  - [x] Update `buildFor` logic
    - [x] #896 
    - [x] #899
    - [x] #901
    - [x] #905
    - [x] #909 
- [x] Migrate `Hashable` implementation for `XCScheme.BuidableReference` to `tuist/XcodeProj`.
  - [x] tuist/XcodeProj#712
  - [x] #912 
- [x] #913 

## Deferred Items
- Be sure that typos in any of the labels specified in the schemes are checked for validity. [Slack thread](https://ios-dev-at-scale.slack.com/archives/C03BS1YT8JE/p1656675962659949)
  - It turns out that `Label()` does not validate that a label is valid (i.e., exists).
  - @brentleyjones and I discussed some options for other ways to validate the existence of a label. Namely, we could add an attribute to the `_xcodeproj` rule that references all of the labels that are specified in the schemes. The downside to this approach is that it may cause analysis of the specified targets using the default (i.e., incorrect) configuration. Also, if a failure occurs, it could be confusing to understand what is causing the error (e.g., `xcodeproj` vs the scheme definition).
  - In short, we are punting on this for now.
- #857

**Comment by BalestraPatrick** [comment]
I'd also add one more feature we'd like to support:
- Specify custom "Argument Passed On Launch" and "Environment Variables".

**Comment by cgrindel** [comment]
I will put together a design proposal for review. I should be able to put up a first pass at the design by the end of the week.

**Comment by brentleyjones** [comment]
FYI @erikkerber since you've expressed interest in this in the past as well.

**Comment by cgrindel** [comment]
@brentleyjones For the rule/macro names, do you want to prefix `xcodeproj_` to the names. For example, for a scheme rule, would you like `xcodeproj_scheme`?  Other options that come to mind: `xcode_scheme`, `scheme`? I, personally, think that `scheme` seems a bit naked and could collide with another rule.

**Comment by brentleyjones** [comment]
People can rename rules when loading, so I'm not concerned with prefixing, and would like `scheme` or `schemes` depending on how the design unfolds.

**Comment by brentleyjones** [comment]
Maybe `xcode_scheme` for clarity though, since someone else reading the build files might not know what a generic "scheme" is.

**Comment by brentleyjones** [comment]
We need to adjust any error messages related to unknown labels, as using an `alias` label will work for project generation, but will fail to resolve to `Target.label`.

**Comment by brentleyjones** [comment]
I believe we need to remove the inferring of top level targets, as it doesn't work with `device_and_simulator`: it results in improperly configured top-level targets being added to `xcodeproj`.

**Comment by brentleyjones** [comment]
The logic for `isTopLevel` is wrong, as we can have top-level targets listed in `build_action`, but nowhere else. For example, a scheme setup like this:
```
xcodeproj(
    name = "xcodeproj",
    project_name = "Multiplatform",
    schemes = [
        xcode_schemes.scheme(
            name = "All",
            build_action = xcode_schemes.build_action([
                "//examples/multiplatform/iMessageApp",
                "//examples/multiplatform/tvOSApp",
                "//examples/multiplatform/Tool",
            ]),
            launch_action = xcode_schemes.launch_action(
                "//examples/multiplatform/iOSApp",
            ),
        ),
    ],
    tags = ["manual"],
    top_level_targets = XCODEPROJ_TARGETS,
)
```

results in this error:
```
No `TargetID` value found for "//examples/multiplatform/tvOSApp:tvOSApp"
```

**Comment by cgrindel** [comment]
> We need to adjust any error messages related to unknown labels, as using an `alias` label will work for project generation, but will fail to resolve to `Target.label`.

@brentleyjones Do you have an example that fails?

**Comment by brentleyjones** [comment]
Create an alias for a top level target, use that alias in the scheme.

**Comment by cgrindel** [comment]
I moved the `top_level_only` task to its own [feature request issue](https://github.com/buildbuddy-io/rules_xcodeproj/issues/857), as it is related to but not required for custom, Xcode schemes.

**Comment by cgrindel** [comment]
Excerpt from Slack conversation about `buildFor` changes.

> Related, we need to change the buildFor logic. Something like this:
buildActionInfo can define which ones (in BUILD files), but defaults to all (like today)
testActionInfo implies testing if buildActionInfo hasn't set any, and an error if buildActionInfo has excluded the testing action
launchActionInfo implies running if buildActionInfo hasn't set any, and an error if buildActionInfo has excluded the running action
profileActionInfo is defaulted to launchActionInfo's target if it's not defined, and if buildActionInfo hasn't excluded the profiling for the target. Then it implies profiling if buildActionInfo hasn't set any, and an error if buildActionInfo has excluded the profiling action
archivingActionInfo is defaulted to launchActionInfo's target if it's not defined, and if buildActionInfo hasn't excluded the archiving for the target. Then it implies archiving if buildActionInfo hasn't set any, and an error if buildActionInfo has excluded the archiving action
Finally, after all of that, if no targets have running set (say a custom scheme with just testActionInfo set), then all targets gain running
Reason: to not build tests when building a custom scheme, unless the test action is taken. This combined with https://ios-dev-at-scale.slack.com/archives/C03JZFP9PHC/p1659390651088049?thread_ts=1659014885.455309&cid=C03JZFP9PHC will prevent the over-building RH is reporting.

**Comment by brentleyjones** [comment]
Nice!

We should open a new issue for the `test_env` request from our Slack discussion.
